### ISSUE TITLE
How did you achieve effecient OpenCL conv gradient computations?

### ISSUE THREAD

**Issue by iperov** [plaidml/plaidml#1451]
I coded my own OpenCL ML library from scratch on pure python with only numpy dependency.
Most ops have the same performance as tensorflow, such as element_wise ops, reductions and other

But conv kernel gradient computation is x14 slower.
Bottleneck is access to large areas of memory, because for every 
IN_CH,OUT_CH,KH,KW of kernel element we have to iterate over all H,W of input and output.

as far as I understand your Tile language is designed to do anything with arrays and get the best performance.

Can you please explain me how did you achieve speed of convs almost the same as tensorflow?

**Comment by jbruestle** [comment]
Honestly the answer is specific both to the architecture of the particular GPU, the details of the convolution (3x3 vs 1x1, striding, dialation, etc), and the memory layout (channels first vs channels last).  The goal of a tensor compiler like PlaidML is to automatically search the entire space of tilings / loops transforms / etc, and pick the best way to do things based on a cost model that considers things like global memory access, alignment, bank conflicts on shared memory, etc.  Even as the person who wrote some of the cost models and has written many kernels by hands, I'm sometimes surprised by the actual generated code, so I don't know the answer off the top of my head.  That said, you can set: PLAIDML_OPENCL_CACHE=<some_dir> to dump the generated OpenCL code and take a look at it.  Also, you can set PLAIDML_STRIPE_OUTPUT=<some_dir> to dump the various intermediate optimizer passes as stripe IR which may be slightly more readable.

**Comment by iperov** [comment]
Everything came down to the speed of a simple matmul...

this is my opencl matmul implementation with local memory
```
{ph.define_axes_accessor('A', a_shape)}
{ph.define_axes_accessor('B', b_shape)}
{ph.define_axes_accessor('O', output_shape)}

#define BLOCK_SIZE {bs}

__kernel void impl(__global float* O, __global const float* A, __global const float* B)
{{
    size_t o0 = get_global_id(0);
    size_t o1 = get_global_id(1);

#if BLOCK_SIZE != 1
    size_t to0 = get_local_id(0);
    size_t to1 = get_local_id(1);
#endif

    float v = 0.0;

    for (size_t a1=0; a1<A1; a1 += BLOCK_SIZE )
    {{
#if BLOCK_SIZE != 1
        __local float As[BLOCK_SIZE][BLOCK_SIZE];
        __local float Bs[BLOCK_SIZE][BLOCK_SIZE];
        
        As[to1][to0] = A[A_idx(o0,a1+to1)];
        Bs[to0][to1] = B[B_idx(o1,a1+to0)];

        barrier(CLK_LOCAL_MEM_FENCE);

        for (size_t k=0; k<BLOCK_SIZE; ++k)
            v += As[k][to0] * Bs[k][to1];

        barrier(CLK_LOCAL_MEM_FENCE);
#else
        v += A[A_idx(o0,a1)] * B[B_idx(o1,a1)];
#endif
    }}
    O[O_idx(o0,o1)] {'+=' if is_add_to_output else '='} v;
}}
```

I am testing it with theese shapes
```
(512, 32768)  x (32768, 4608)
```

vs tf matmul
```
t1 = tf.get_variable ("t1", (512,32768), dtype=tf.float32)
t2 = tf.get_variable ("t2", (32768,4608), dtype=tf.float32)
loss = tf.matmul(t1,t2)
```
my speed = 225ms

first iteration of tensorflow is 210ms, but next is 13ms. Thus, TF is 16x faster.

is this some CUDA magic?

**Comment by iperov** [comment]
hmm
I tested it in plaidml 0.7.0

```
os.environ['PLAIDML_EXPERIMENTAL'] = 'false'
os.environ["KERAS_BACKEND"] = "plaidml.keras.backend"
os.environ["PLAIDML_DEVICE_IDS"] = "opencl_nvidia_geforce_rtx_2080_ti.0"
    
import keras
import plaidml    
import plaidml.tile
PML = plaidml
PMLK = plaidml.keras.backend
PMLTile = plaidml.tile
    
t1 = PMLK.placeholder( (512,32768) )
t2 = PMLK.placeholder( (32768,4608) )
d = PMLK.dot(t1,t2)
f = PMLK.function( [t1,t2], [d ])
    
for i in range(100):
    n1 = np.random.randint( 2**8, size=(512,32768) ).astype(np.float32)
    n2 = np.random.randint( 2**8, size=(32768,4608) ).astype(np.float32)
    t = time.time()
    x = f([n1,n2])
    print(f'time {time.time()-t} ')
```
result is 567ms. 
So my opencl implementation is faster :D
